<?xml version="1.0" encoding="utf-8"?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from manipulator.urdf.xacro         | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="manipulator_complete">
  <material name="green">
    <color rgba="0 1 0 1"/>
  </material>
  <material name="black">
    <color rgba="0.4 0.4 0.4 1"/>
  </material>
  <material name="yellow">
    <color rgba="247 202 24 1"/>
  </material>
  <gazebo reference="base_link">
    <material>Gazebo/Yellow</material>
  </gazebo>
  <gazebo reference="top_base">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="front_left_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="front_right_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="rear_left_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="rear_right_wheel">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="arm_base">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="bicep">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="bottom_wrist">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="elbow">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="top_wrist">
    <material>Gazebo/Black</material>
  </gazebo>

  <!--ROS CONTROL-->
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <robotNamespace>/</robotNamespace>
      <controlPeriod>0.001</controlPeriod>
      <legacyModeNS>false</legacyModeNS>
    </plugin>
  </gazebo>
  <!--DIFF DRIVE -->
  <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="diff_drive_controller">
      <legacyMode>false</legacyMode>
      <alwaysOn>true</alwaysOn>
      <updateRate>1000.0</updateRate>
      <leftJoint>front_left_wheel_joint, rear_left_wheel_joint</leftJoint>
      <rightJoint>front_right_wheel_joint, rear_right_wheel_joint</rightJoint>
      <wheelSeparation>0.72</wheelSeparation>
      <wheelDiameter>0.4</wheelDiameter>
      <wheelTorque>10</wheelTorque>
      <publishTf>1</publishTf>
      <odometryFrame>map</odometryFrame>
      <commandTopic>cmd_vel</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <robotBaseFrame>base_link</robotBaseFrame>
      <wheelAcceleration>2.8</wheelAcceleration>
      <publishWheelJointState>true</publishWheelJointState>
      <publishWheelTF>false</publishWheelTF>
      <odometrySource>world</odometrySource>
      <rosDebugLevel>Debug</rosDebugLevel>
    </plugin>
  </gazebo>
  <!-- BASE LINK -->
  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/base.stl" scale="0.15 0.1 0.1"/>
      </geometry>
      <origin rpy="0 0 1.57" xyz="0 0 0.12"/>
      <material name="green"/>
    </visual>
    <collision>
      <origin rpy="0 0 1.57" xyz="0 0 0.12"/>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/base.stl" scale="0.15 0.1 0.1"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.12"/>
      <mass value="50"/>
      <inertia ixx="1.2" ixy="-1.03781944357671E-10" ixz="0.00763014265820928" iyy="1.9" iyz="1.09578155845563E-12" izz="2.5"/>
    </inertial>
  </link>
  <!-- TOP PLATE -->
  <link name="top_base">
    <visual>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/top_base.stl" scale="0.15 0.1 0.1"/>
      </geometry>
      <origin xyz="0 0 0"/>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/top_base.stl" scale="0.15 0.1 0.1"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1"/>
      <inertia ixx="0.00856502765719703" ixy="1.5074118157338E-19" ixz="-4.78150098725052E-19" iyy="0.013670640432096" iyz="-2.68136447099727E-19" izz="0.00856502765719703"/>
    </inertial>
  </link>
  <joint name="base_link_to_top_base_link" type="fixed">
    <parent link="base_link"/>
    <child link="top_base"/>
    <origin rpy="0 0 0" xyz="0 0 0.26"/>
  </joint>
  <!--WHEEL LINK-->
  <link name="front_left_wheel">
    <visual>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
    </visual>
    <collision>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="2.5"/>
      <inertia ixx="0.00856502765719703" ixy="1.5074118157338E-19" ixz="-4.78150098725052E-19" iyy="0.013670640432096" iyz="-2.68136447099727E-19" izz="0.00856502765719703"/>
    </inertial>
  </link>
  <link name="front_right_wheel">
    <visual>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
    </visual>
    <collision>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="2.5"/>
      <inertia ixx="0.00856502765719703" ixy="1.5074118157338E-19" ixz="-4.78150098725052E-19" iyy="0.013670640432096" iyz="-2.68136447099727E-19" izz="0.00856502765719703"/>
    </inertial>
  </link>
  <link name="rear_left_wheel">
    <visual>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
    </visual>
    <collision>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="2.5"/>
      <inertia ixx="0.00856502765719703" ixy="1.5074118157338E-19" ixz="-4.78150098725052E-19" iyy="0.013670640432096" iyz="-2.68136447099727E-19" izz="0.00856502765719703"/>
    </inertial>
  </link>
  <link name="rear_right_wheel">
    <visual>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
    </visual>
    <collision>
      <origin rpy="1.57 0 -1.57" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://mobile_manipulator/meshes/left_wheel.stl" scale="0.15 0.15 0.15"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="2.5"/>
      <inertia ixx="0.00856502765719703" ixy="1.5074118157338E-19" ixz="-4.78150098725052E-19" iyy="0.013670640432096" iyz="-2.68136447099727E-19" izz="0.00856502765719703"/>
    </inertial>
  </link>
  <!--WHEEL JOINT-->
  <joint name="front_left_wheel_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="base_link"/>
    <child link="front_left_wheel"/>
    <origin rpy="0 0 0" xyz="0.220 0.36 0"/>
  </joint>
  <joint name="front_right_wheel_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="base_link"/>
    <child link="front_right_wheel"/>
    <origin rpy="0 0 0" xyz="0.220 -0.36 0"/>
  </joint>
  <joint name="rear_left_wheel_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="base_link"/>
    <child link="rear_left_wheel"/>
    <origin rpy="0 0 0" xyz="-0.220 0.36 0"/>
  </joint>
  <joint name="rear_right_wheel_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="base_link"/>
    <child link="rear_right_wheel"/>
    <origin rpy="0 0 0" xyz="-0.220 -0.36 0"/>
  </joint>
  <!--WHEEL TRANSMISSION-->
  <transmission name="front_left_wheel_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="front_left_wheel_motor">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="front_left_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="front_right_wheel_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="front_right_wheel_motor">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="front_right_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="rear_left_wheel_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="rear_left_wheel_motor">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rear_left_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="rear_right_wheel_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="rear_right_wheel_motor">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rear_right_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <!-- WHEEL FRICTION -->
  <gazebo reference="front_left_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <gazebo reference="front_right_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <gazebo reference="rear_left_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <gazebo reference="rear_right_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <!--######################################################################-->
  <!--######################################################################-->
  <!--################################ ROS-Controllers ######################################-->
  <gazebo>
    <plugin filename="libgazebo_ros_joint_state_publisher.so" name="joint_state_publisher">
      <jointName>arm_base_joint, shoulder_joint, bottom_wrist_joint, elbow_joint, bottom_wrist_joint</jointName>
    </plugin>
  </gazebo>
  <!--######################################################################-->
  <link name="arm_base">
    <inertial>
      <origin rpy="0 0 0" xyz="7.7128E-09 -0.063005 -3.01969999961422E-08"/>
      <mass value="1.6004"/>
      <inertia ixx="0.00552196561445819" ixy="7.9550614501301E-10" ixz="-1.34378458924839E-09" iyy="0.00352397447953875" iyz="-1.10071809773382E-08" izz="0.00553739792746489"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/arm_base.stl"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/arm_base.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="bicep">
    <inertial>
      <origin rpy="0 0 0" xyz="0.12821 3.5589E-06 0.052492"/>
      <mass value="1.1198"/>
      <inertia ixx="0.0012474" ixy="-5.4004E-07" ixz="-0.0013148" iyy="0.0072923" iyz="-1.8586E-07" izz="0.0068178"/>
    </inertial>
    <visual>
      <origin rpy="0 -1.5708 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/bicep.stl"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 -1.5708 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/bicep.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="bottom_wrist">
    <inertial>
      <origin rpy="0 0 0" xyz="-9.1053E-08 -0.069257 -1.86629999995759E-07"/>
      <mass value="0.27721"/>
      <inertia ixx="0.00104290750143942" ixy="4.37155302268076E-09" ixz="-2.45049603914627E-09" iyy="0.000380518373895034" iyz="-7.56009835172156E-09" izz="0.00106006525067445"/>
    </inertial>
    <visual>
      <origin rpy="3.14 0 1.5708" xyz="0 0 0.13522"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/wrist.stl"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="3.14 0 1.5708" xyz="0 0 0.13522"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/wrist.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="elbow">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.11109 1.1476E-08 0.046469"/>
      <mass value="0.84845"/>
      <inertia ixx="0.00079656" ixy="-7.8011E-10" ixz="0.00053616" iyy="0.003576" iyz="4.6326E-10" izz="0.0033698"/>
    </inertial>
    <visual>
      <origin rpy="0 -1.5708 0" xyz="0.0 0.05163 0.20994"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/elbow.stl"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 -1.5708 0" xyz="0.0 0.05163 0.20994"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/elbow.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="top_wrist">
    <inertial>
      <origin rpy="0 0 0" xyz="-9.1053E-08 -0.069257 -1.86629999995759E-07"/>
      <mass value="0.27721"/>
      <inertia ixx="0.00104290750143942" ixy="4.37155302268076E-09" ixz="-2.45049603914627E-09" iyy="0.000380518373895034" iyz="-7.56009835172156E-09" izz="0.00106006525067445"/>
    </inertial>
    <visual>
      <origin rpy="3.14 0 1.5708" xyz="0 0 0.13522"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/wrist.stl"/>
      </geometry>
      <material name="">
        <color rgba="0.79216 0.81961 0.93333 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="3.14 0 1.5708" xyz="0 0 0.13522"/>
      <geometry>
        <mesh filename="package://robot_description/meshes/wrist.stl"/>
      </geometry>
    </collision>
  </link>
  <!--######################################################################-->
  <joint name="arm_base_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="base_link"/>
    <child link="arm_base"/>
    <origin rpy="0 0 1.5708" xyz="0.2 0 0.265"/>
  </joint>
  <joint name="shoulder_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="arm_base"/>
    <child link="bicep"/>
    <origin rpy="0 0 1.5708" xyz="-0.05166 0.0 0.20271"/>
  </joint>
  <joint name="bottom_wrist_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="bicep"/>
    <child link="bottom_wrist"/>
    <origin rpy="0 0 0" xyz="0.0 -0.05194 0.269"/>
  </joint>
  <joint name="elbow_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="bottom_wrist"/>
    <child link="elbow"/>
    <origin rpy="0 0 0" xyz="0.0 0 0.13522"/>
  </joint>
  <joint name="top_wrist_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <parent link="elbow"/>
    <child link="top_wrist"/>
    <origin rpy="0 0 0" xyz="0.0 0 0.20994"/>
  </joint>
  <!--######################################################################-->
  <transmission name="arm_base_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="arm_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="arm_base_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="shoulder_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="shoulder_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="shoulder_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="bottom_wrist_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="bottom_wrist_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="bottom_wrist_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="elbow_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="elbow_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="elbow_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <transmission name="top_wrist_trans" type="SimpleTransmission">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="top_wrist_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="top_wrist_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  <!--######################################################################-->
</robot>

